airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/state/StateManager.kt

```diff
@@ -44,50 +39,73 @@ interface StateManager {
  * TODO: Ensure that state is flushed at the end, and require that all state be flushed before the
  * destination can succeed.
  */
-@Singleton
-class DefaultStateManager(
-    private val catalog: DestinationCatalog,
-    private val streamsManager: StreamsManager,
-    private val stateMessageFactory: AirbyteStateMessageFactory,
-    private val outputConsumer: OutputConsumer
-) : StateManager {
+abstract class StreamsStateManager<T, U>() : StateManager<DestinationStream, T> {
     private val log = KotlinLogging.logger {}
 
-    data class GlobalState(
+    abstract val catalog: DestinationCatalog
+    abstract val streamsManager: StreamsManager
+    abstract val outputFactory: MessageConverter<T, U>
+    abstract val outputConsumer: Consumer<U>
+
+    data class GlobalState<T>(
         val streamIndexes: List<Pair<DestinationStream, Long>>,
-        val stateMessage: DestinationStateMessage
+        val stateMessage: T
     )
 
     private val stateIsGlobal: AtomicReference<Boolean?> = AtomicReference(null)
```

Comment:
The reason it's an atomic reference is because setting it `null -> true` and `null -> false` is also an operation that needs to be guarded. I think. (If stateIsGlobal needs to be then it needs to be.)